### PR TITLE
[TASK] refactor content wizard include

### PR DESCRIPTION
**Pull request by kjeldschumacher**
Hi,
I have changed the following things:
1. removed content wizard adding functions inside class tx_multicolumn_wizardItemsHook
2. fixed misspelling of multiColumnParentId inside class tx_multicolumn_wizardItemsHook
3. changed intval to int for GeneralUtility::_GP('tx_multicolumn_parentid')
4. added Configuration/TSconfig folder
5. moved tsconfig.txt to Configuration/TSconfig/page.ts
6. added wizard configuration to Configuration/TSconfig/ContentWizard/_wizard.ts

### CODE REVIEW

File: Configuration/TSconfig/ContentWizard/_wizard.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# add multicolumn container to common tab
```
Comment (IchHabRecht) [inline Configuration/TSconfig/ContentWizard/_wizard.ts]: Please name the file NewContentElementWizard.ts (in TSconfig folder) and include it separately in ext_localconf.php
Please remove the comment in the first line.

File: Configuration/TSconfig/ContentWizard/_wizard.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+mod {
```
Comment (IchHabRecht) [inline Configuration/TSconfig/ContentWizard/_wizard.ts]: Please prevent unnecessary brackets.
mod.wizards.newContentElement.wizardItems.common.elements should be the first line

File: Configuration/TSconfig/page.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
+# include contentWizard settings
+<INCLUDE_TYPOSCRIPT: source="FILE:./ContentWizard/_wizard.ts">
```
Comment (IchHabRecht) [inline Configuration/TSconfig/page.ts]: Please include the TSconfig file in ext_localconf.php

File: Configuration/TSconfig/page.ts [hunk 4]
```diff
@@ -0,0 +1,299 @@
+# include contentWizard settings
```
Comment (IchHabRecht) [inline Configuration/TSconfig/page.ts]: Please name the file multicolumn.ts as there is not page but mutlicolumn configuration.

File: ext_localconf.php [hunk 5]
```diff
@@ -22,7 +22,8 @@
 $TYPO3_CONF_VARS['SC_OPTIONS']['tce']['formevals']['tx_multicolumn_tce_eval'] = '';
 
 //add page TSconfig
-\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addPageTSConfig('<INCLUDE_TYPOSCRIPT: source="FILE:EXT:multicolumn/tsconfig.txt">');
```
Comment (IchHabRecht) [inline ext_localconf.php]: Include both .ts files here

File: Configuration/TSconfig/ContentWizard/_wizard.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    wizards.newContentElement.wizardItems.common {
+        elements {
+            multicolumn {
+                iconIdentifier = tx-multicolumn-wizard-icon
```
Comment (IchHabRecht) [inline Configuration/TSconfig/ContentWizard/_wizard.ts]: There is no need for an own identifier. Please use EXT:multicolum/...

File: ext_tables.php [hunk 7]
```diff
@@ -13,6 +13,14 @@
     $GLOBALS['TYPO3_CONF_VARS']['BE']['ContextMenu']['ItemProviders'][1492078309] = \CPSIT\Multicolumn\ContextMenu\ItemProvider::class;
 }
 
+// register wizard icon
```
Comment (IchHabRecht) [inline ext_tables.php]: Please remove the identifier registration.

File: hooks/class.tx_multicolumn_wizardItemsHook.php [hunk 8]
```diff
@@ -22,104 +22,45 @@
  *
  *  This copyright notice MUST APPEAR in all copies of the script!
  ***************************************************************/
+
+use TYPO3\CMS\Core\Utility\GeneralUtility;
```
Comment (IchHabRecht) [inline hooks/class.tx_multicolumn_wizardItemsHook.php]: I think the whole hook can be removed. The necessary parent information should come from "new content"-link and is passed to the content element automatically (like the colPos information for content elements).

**Comment by IchHabRecht** [comment]
Hi Kjeld,

I went through the code and it seems it is not possible to pass the tx_multicolumn_parentid value automatically. Means the Hook has to stay in a limited way. Please move it to Classes/Hooks/WizardItemsHook.php. Code should find the correct item and append the tx_multicolumn_parentid from Get/Post vars.

Thanks for your help.

File: Classes/Hooks/WizardItemsHook.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+use TYPO3\CMS\Backend\Wizard\NewContentElementWizardHookInterface;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+class WizardItemsHook implements NewContentElementWizardHookInterface
+{
+    /**
+     * @var int
+     */
+    protected $multiColumnParentId;
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: No need for an own class property here.

File: Classes/Hooks/WizardItemsHook.php [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+     * modifies WizardItems array
+     *
+     * @param array $wizardItems array of Wizard Items
+     * @param \TYPO3\CMS\Backend\Controller\ContentElement\NewContentElementController $parentObject New Content element wizard
+     *
+     * @return    void
+     */
+    public function manipulateWizardItems(&$wizardItems, &$parentObject)
+    {
+        $this->multiColumnParentId = (int)GeneralUtility::_GP('tx_multicolumn_parentid');
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: please just use a local variable to store the parent information.

File: Classes/Hooks/WizardItemsHook.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        if ($this->multiColumnParentId > 0) {
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: if (empty($multiColumnParentId)) {
    return;
}

File: Classes/Hooks/WizardItemsHook.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+            $this->addMulticolumnParentId($wizardItems);
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: No need for a separate function. Please move the code here.

File: Classes/Hooks/WizardItemsHook.php [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+     * @param    array                    array of Wizard Items
+     *
+     * @return    void
+     */
+    protected function addMulticolumnParentId(array &$wizardItems)
+    {
+        foreach ($wizardItems as &$wizardItem) {
+            if ($wizardItem['params']) {
+                //add mulitcolumn parent id
+                $wizardItem['params'] .= '&defVals[tt_content][tx_multicolumn_parentid]=' . $this->multiColumnParentId;
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: Even if there aren't any parameters set yet, the parent information has to be set. Please add it to $wizardItem['tt_content_defValues'] and generate $wizardItem['params'] with GeneralUtility::implodeArrayForUrl('', $wizardItem['tt_content_defValues']);

File: Classes/Hooks/WizardItemsHook.php [hunk 14]
```diff
@@ -0,0 +1,33 @@
+<?php
+namespace CPSIT\Multicolumn\Hooks;
+
+use TYPO3\CMS\Backend\Wizard\NewContentElementWizardHookInterface;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+class WizardItemsHook implements NewContentElementWizardHookInterface
+{
+
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: Superfluous empty line - please remove

File: Classes/Hooks/WizardItemsHook.php [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+     * @return    void
+     */
+    public function manipulateWizardItems(&$wizardItems, &$parentObject)
+    {
+        $multiColumnParentId = (int)GeneralUtility::_GP('tx_multicolumn_parentid');
+        if (empty($multiColumnParentId)) {
+            return;
+        }
+
+        foreach ($wizardItems as &$wizardItem) {
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: This still isn't working. Please fix to create a corrent link using defVals[tt_content] for params attribute.

Please test end ensure working functionality before pushing next time.

File: lib/class.tx_multicolumn_db.php [hunk 16]
```diff
@@ -79,6 +80,10 @@ public static function getContentElementsFromContainer($colPos, $pid, $mulitColu
             if ($isWorkspace) {
                 $queryBuilder->getRestrictions()->add(BackendWorkspaceRestriction::class);
             }
+            if ($showHidden) {
+                $queryBuilder->getRestrictions()->removeByType(HiddenRestriction::class);
```
Comment (IchHabRecht) [inline lib/class.tx_multicolumn_db.php]: As far as I can see, this fix isn't related to the topic. Please remove this change and submit a new issue/pull request. Thank you.

Comment (kpurrmann) [inline lib/class.tx_multicolumn_db.php]: Add #38 and removed it.

File: lib/class.tx_multicolumn_db.php [hunk 17]
```diff
@@ -79,6 +79,7 @@ public static function getContentElementsFromContainer($colPos, $pid, $mulitColu
             if ($isWorkspace) {
                 $queryBuilder->getRestrictions()->add(BackendWorkspaceRestriction::class);
             }
+
```
Comment (IchHabRecht) [inline lib/class.tx_multicolumn_db.php]: As far as I can see, this change isn't needed?

File: Classes/Hooks/WizardItemsHook.php [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            if (empty($wizardItem['tt_content_defValues'])) {
+                $wizardItem['tt_content_defValues'] = [];
+            }
+            $wizardItem['tt_content_defValues']['tx_multicolumn_parentid'] = $multiColumnParentId;
+            $wizardItem['params'] = GeneralUtility::implodeArrayForUrl('defVals[tt_content]', $wizardItem['tt_content_defValues']);
```
Comment (IchHabRecht) [inline Classes/Hooks/WizardItemsHook.php]: This adds the params parameter for tab items as well. would you mind to exclude those from the processing:

            if (strpos($key, '_') === false) {
                continue;
            }

**Pull request merged**